utilities/transactions/pessimistic_transaction_db.cc

```diff
@@ -280,22 +288,32 @@ Status TransactionDB::WrapStackableDB(
     StackableDB* db, const TransactionDBOptions& txn_db_options,
     const std::vector<size_t>& compaction_enabled_cf_indices,
     const std::vector<ColumnFamilyHandle*>& handles, TransactionDB** dbptr) {
-  PessimisticTransactionDB* txn_db;
+  assert(db != nullptr);
+  assert(dbptr != nullptr);
+  *dbptr = nullptr;
+
+  std::unique_ptr<StackableDB> root(db);
+  std::unique_ptr<PessimisticTransactionDB> txn_db;
+
   switch (txn_db_options.write_policy) {
     case WRITE_UNPREPARED:
       return Status::NotSupported("WRITE_UNPREPARED is not implemented yet");
     case WRITE_PREPARED:
-      txn_db = new WritePreparedTxnDB(
-          db, PessimisticTransactionDB::ValidateTxnDBOptions(txn_db_options));
+      txn_db.reset(new WritePreparedTxnDB(
+          root.release(), PessimisticTransactionDB::ValidateTxnDBOptions(txn_db_options)));
       break;
     case WRITE_COMMITTED:
     default:
-      txn_db = new WriteCommittedTxnDB(
-          db, PessimisticTransactionDB::ValidateTxnDBOptions(txn_db_options));
+      txn_db.reset(new WriteCommittedTxnDB(
+          root.release(), PessimisticTransactionDB::ValidateTxnDBOptions(txn_db_options)));
   }
   txn_db->UpdateCFComparatorMap(handles);
-  *dbptr = txn_db;
   Status s = txn_db->Initialize(compaction_enabled_cf_indices, handles);
+  // In case of failure db is deleted along with the
+  // tx database
```

Comment:
@yuslepukhin The semantics I suggested seems quite reasonable to me: 
- ::WrapDB would not delete the passed root db if the wrapping failed. Upon failure it would cleanup any temp object that it created though.
- ::Open would delete the root db that it created if the subsequent WrapDB failed

I think this would work without the need of any API change. API change should be used rarely only when it is the last resort. My interest in this issue is to fix up the pessimistic transactions. The improvement on optimistic transaction could be left for other PRs.

Reply:
@maysamyabandeh Here is a question to make this understand better. You are saying that txn_db does not delete root db. Meaning, it does not own it. So who owns it when it is successful? Most people would assume that one ptr they deal with the result is what they have to cleanup.

Reply:
Based on the current code status (upon which applications has been built) whoever owns txn_db, owns the root db as well. So the application should explicitly delete the root db after they delete the txn_db. We are at a stage that we do not have much freedom on changing the contract of public functions. The best we can do is to clarify the contract and document it.

Reply:
@maysamyabandeh Sure, I am not going to impose my opinion on you. However, your test code leaks on failure and provides a wrong example. Also it is not consistent and is not well documented. Thanks for your consideration.

Reply:
Certainly! We are working together to make rocksdb better. So should we update this patch based on the above semantics that we agreed upon?
> your test code leaks on failure
Is there any other leak that is not fixed by your patch?
> provides a wrong example
Can you point me to the example?
> it is not consistent
I guess there might be other inconsistency issues too. We can improve them either as part of this patch or later patches.
> not well documented
Yeap that is a major issue. We try to improve the documentation with each new patch.

Reply:
You can see that `new StackableDB()` leaks right away and root_db leaks with it. Technically, you can delete root_db but only if you keep leaking StackableDB because it owns it. Inconsistencies:
- DBImpl:;Open requires no cleanup on failure but TransactionDB does.
- StackableDB() owns the database that is wrapping but your Open/Wrap does not

All of that is not clearly documented. A person who makes use of this API must keep both pointers and destroy them in the correct order regardless of failure or success.

In the example above, according to your description I am supposed to destroy db. Problem is that on some failures db is populated and on another failure it is not populated but the test does not always init db to nullptr on closure so there is not a way to know. Why not to make it clear and consistent? If there is a failure, there is nothing to cleanup. If it is a success then one clearly gets back a database object that one is responsible to cleanup. 

I am sorry, that I was not clear before and may be even unclear now. However, I feel I have spent enough time on it.

Reply:
I totally understand.; do not worry about it. I will take it from here and apply the suggested improvements. Thanks for your contribution.

Reply:
Correction: TransactionDB takes ownership of the root db. It was my bad missing the StackableDB destructor.